ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1129,6 +1267,200 @@ private void applyDelete(MessageInfo messageInfo, RemoteReplicaInfo remoteReplic
+            + " and replicasToReplicatePerNode size " + remoteReplicaInfos.size() + " should be the same");
+      }
+
+      for (int i = 0; i < remoteReplicaInfos.size(); i++) {
+        RemoteReplicaInfo remoteReplicaInfo = remoteReplicaInfos.get(i);
+        String partitionName = remoteReplicaInfo.getLocalReplicaId().getPartitionId().toPathString();
+        ReplicaId remoteReplicaId = remoteReplicaInfo.getReplicaId();
+
+        // Check if local replica and remote replica are leaders for this partition.
+        if (leaderBasedReplicationAdmin.isPeerReplicaLeaderForPartition(partitionName, remoteReplicaId)) {
+          leaderReplicaInfosOutput.add(remoteReplicaInfo);
+          exchangeMetadataResponseListForLeaderReplicaInfosOutput.add(exchangeMetadataResponseList.get(i));
+        }
+      }
+    } else {
+      // if leader based replication is disabled or we are replicating within intra-colo, include all remote replicas for
+      // replication.
+      leaderReplicaInfosOutput.addAll(remoteReplicaInfos);
+      exchangeMetadataResponseListForLeaderReplicaInfosOutput.addAll(exchangeMetadataResponseList);
+    }
+  }
+
+  /**
+   * Returns list of remote replica infos from a given remote node whose missing blobs in their metadata response
+   * haven't arrived within time = replicationConfig.replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds.
+   *
+   * @param remoteReplicaInfos list of remote replica infos
+   * @return list of remote replica infos which have timed out due to no progress
+   */
+  private List<RemoteReplicaInfo> getRemoteStandbyReplicasTimedOutOnNoProgress(
+      List<RemoteReplicaInfo> remoteReplicaInfos) {
+
+    // Use case: In leader-based replication, standby replicas don't send replication GET requests for missing keys
+    // found in their metadata exchange and expect them to come from leader in local data center via intra-dc replication.
+    // However, if for any reason, their missing blobs never arrive via local leader, this is a safety feature to fetch
+    // the blobs themselves in order to avoid being stuck.
+
+    // Example scenario: For DELETE after PUT use case in remote data center, it is possible that standby replicas get
+    // only PUT record in its replication cycle (DELETE record will come in next cycle) while leader gets both
+    // PUT and DELETE together in its replication cycle. Due to that, leader doesn't fetch
+    // the deleted blob from remote data center and the blob is never replicated from leader to standby.
+    // As a result, the PUT record in standby's missing blobs set is never emptied.
+
+    // Time out period is configurable via replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds. If
+    // replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds == -1, this safety feature is disabled.
+
+    List<RemoteReplicaInfo> remoteReplicasTimedOut = new ArrayList<>();
+    if (replicatingFromRemoteColo && leaderBasedReplicationAdmin != null
```

Comment:
nit: the repilcatingFromRemoteColo and leaderBaseReplicationAdmin check seem unnecessary, for this method only used when they are true.